jep/0000/README.adoc

```diff
@@ -0,0 +1,220 @@
+plugin developers to document required breaking changes, and validate a configuration for a specific
+jenkins instance before it get actually applied.
+
+== Reasoning
+
+Without an explicit versioned model per jenkins-core and plugin versions Configuration-as-Code only
+can rely offer some minimal backward compatibility.
+
+
+=== Relax Configuration-asCode for backward compatibility ?
+
+There's few circumstances where we have options
+to workaround a technical change in Jenkins or plugin codebase as the technical bits are still available,
+just hidden by the default "model inference" mechanism.
+
+A typical sample for this has been debated on
+https://github.com/jenkinsci/configuration-as-code-plugin/issues/233[issue 233]: `JNLPLauncher` DataBound
+constructor has been updated in jenkins codebase to introduce new argument `workDirSettings`. This one
+is declared in API as `@Nonnull`, which means Configuration-as-Code will infer a required attribute. As
+a result, previous definition for a JNLP launcher will be broken as this attribute isn't declared. But on
+the other side, the legacy constructor is still available (but deprecated) and do define a default value
+for this new argument.
+
+**Shall we just relax Configuration-as-Code so it still can apply an outdated model ?**
+
+We don't think so, as doing so would eventually bring us to indeterminable configuration paths:
+As we allow use of deprecated constructor (or setters, or ...) we might result with colliding candidates
+to invoke, without a deterministic way to select the adequate one.
+
+To follow-up with this specific sample, JNLPLauncher do define (at time writing) two deprecated constructors
+and another non-deprecated one :
+```java
+@Deprecated
+public JNLPLauncher(String tunnel, String vmargs) { ... }
+
+@Deprecated
+public JNLPLauncher() { ... }
+
+public JNLPLauncher(boolean enableWorkDir) { ... }
+
+```
+The sole one we can guarantee is designed to submit configuration is the @DataBound one. For others
+there's no deterministic way to choose one. Even considering the configuration attributes we have from
+yaml file, which one should we use if configuration doesn't define `enableWorkDir`, `tunnel` nor `vmargs`?
+first one passing `null, null` of non-deprecated one passing `false` default value ?
+
+=== During Jenkins development
+
+Actually the exact same code change discussed in https://github.com/jenkinsci/configuration-as-code-plugin/issues/233[issue 233]
+could have been implemented without breaking Configuration-as-Code inferred model :
+
+- by defining a `@DataBoundSetter` for newly introduced `workDirSettings` attribute
+- by not defining `workDirSettings` argument as `@Nonnull` and applying default value
+
+Assuming Configuration-as-Code would exist for some time when this change was applied to jenkins-core, we
+might have detected it during code review and prevented this to happen.
```

Comment:
In fact this change to `JNLPLauncher` was wrong and should be fixed: if there was a reasonable default for `workDirSettings`, then it should have been in a `@DataBoundSetter` and thus optional. And there _is_ such a [reasonable default](https://github.com/jenkinsci/jenkins/blob/d4737b2b60438ce0e3cb7a8a6f26675f7c880914/core/src/main/java/jenkins/slaves/RemotingWorkDirSettings.java#L55) as seen also [here](https://github.com/jenkinsci/jenkins/blob/d4737b2b60438ce0e3cb7a8a6f26675f7c880914/core/src/main/resources/jenkins/slaves/RemotingWorkDirSettings/config.jelly#L28-L37). CC @oleg-nenashev